### PR TITLE
Project 5 - Sara Heijkenskjöld

### PR DESCRIPTION
**Pull request by sarahejhej**
This week I felt I got a bit lost in my own code. There are three things I really would like feedback on:
- I tried to write some universal functions, but I'm not sure if they made things clearer or messier?
- I added event listeners for the first time but I'm not sure where in the js file I should put them? For now they're at the end of the file.
- I've made a bunch of tiny, more simple functions instead of one giant, complex function. I don't know if that's a good idea or bad practice?

Lastly, the formatting is a bit weird in both the HTML and the js files. I tried to change my formatting settings, but I didn't get it to worked. The files were like that when I forked them.

### CODE REVIEW

**Comment by sarahejhej** [comment]
Link to live webpage: https://wonderful-shannon.netlify.app/

File: code/index.html [hunk 1]
```diff
@@ -5,33 +5,82 @@
     <meta name="viewport" content="width=device-width, initial-scale=1.0" />
     <meta http-equiv="X-UA-Compatible" content="ie=edge" />
     <link rel="stylesheet" href="style.css" />
```
Comment (trangtara) [inline code/index.html:7]: i just learned that we should add './' before the 'style.css' just to make it consistent while adding link. Good habit to start with. :-)

File: code/index.html [hunk 2]
```diff
@@ -5,33 +5,82 @@
...
-    <title>Technigo Project Cluedo</title>
+    <title>Cluedo</title>
+    <link rel="shortcut icon" href="./assets/splash.png" type="image/png" />
```
Comment (trangtara) [inline code/index.html:9]: You did it great here with adding './' to the path of the image.

File: code/index.html [hunk 3]
```diff
@@ -5,33 +5,82 @@
...
   </head>
   <body>
-    <section class="cards">
-      <article class="killer-deck" id="killerCard">
```
Comment (trangtara) [inline code/index.html:12]: Great practice of naming class and id!!!

File: code/index.html [hunk 4]
```diff
@@ -5,33 +5,82 @@
...
-        <p>KILLER</p>
-        <p id="killerName"></p>
-        <p id="killerAge"></p>
-        <p id="killerOccupation"></p>
-        <img id="killerImage" src="" height="100" />
```
Comment (trangtara) [inline code/index.html:17]: I did the same with you while putting the height attribute here. Some suggestion is to move it to css file instead. I am not really sure which is better or even they make any difference.

File: code/index.html [hunk 5]
```diff
@@ -5,33 +5,82 @@
...
-      </article>
-    </section>
+    <!-- HEADER -->
+    <header class="header-wrapper visible" id="startWindow">
+      <div class="header-container">
+        <h1 class="header-text">CLUEDO</h1>
+        <div class="drop"></div>
+      </div>
+      <button class="button button-start" id="startButton">LET'S PLAY</button>
+    </header>
```
Comment (trangtara) [inline code/index.html:19]: snygg!!!

File: code/index.html [hunk 6]
```diff
@@ -5,33 +5,82 @@
...
 
-    <h1 id="mystery"></h1>
-
-    <button>REVEAL THE CRIME</button>
+    <!-- GAME WINDOW -->
```
Comment (trangtara) [inline code/index.html:21]: I like the fact that you add COMMENT when starting a new section, which makes it much easier to read your code. Nice!!!

File: code/index.html [hunk 7]
```diff
@@ -5,33 +5,82 @@
...
+      <h1 class="mystery-reveal" id="mystery"></h1>
+      <article class="reveal-deck" id="mysteryCard">
+        <p class="p-header">MYSTERY REVEALED</p>
+        <p id="mysteryKiller"></p>
+        <img id="mysteryKillerImage" class="killer-image" src="" />
+        <p id="mysteryRoom"></p>
+        <p id="mysteryWeapon"></p>
+      </article>
+      <div>
+        <button class="button" id="playAgain">PLAY AGAIN</button>
```
Comment (trangtara) [inline code/index.html:81]: smart!!!

File: code/style.css [hunk 8]
```diff
@@ -1,58 +1,377 @@
...
+* {
   margin: 0;
-  font-family: "Courier New", Courier, monospace;
+  box-sizing: border-box;
+}
+
+/*** GENERAL ***/
+
+body {
+  font-family: 'Courier New', Courier, monospace;
```
Comment (trangtara) [inline code/style.css:9]: I see that font-family is already put inside the previous block. We can remove it from the body styling.

File: code/style.css [hunk 9]
```diff
@@ -1,58 +1,377 @@
...
 
-.cards {
-  margin: 50px auto;
+.visible {
+  display: flex;
+}
+
+.invisible {
+  display: none;
+}
```
Comment (trangtara) [inline code/style.css:26]: Super!!!

File: code/style.css [hunk 10]
```diff
@@ -1,58 +1,377 @@
...
+
+/*** HEADER WINDOW ***/
+
+.header-wrapper {
+  flex-direction: column;
+  justify-content: center;
+  height: 100vh;
+}
```
Comment (trangtara) [inline code/style.css:34]: I think we need to have display: flex before activating flex-direction and justify-content.

File: code/style.css [hunk 11]
```diff
@@ -1,58 +1,377 @@
...
+  content: '';
+  position: absolute;
+  width: 0;
+  height: 0;
+  border-left: 10px solid transparent;
+  border-right: 10px solid transparent;
+  border-bottom: 30px solid crimson;
+  top: -22px;
+}
+
```
Comment (trangtara) [inline code/style.css:77]: Wohooooo!!!!

File: code/style.css [hunk 12]
```diff
@@ -1,58 +1,377 @@
...
+  -moz-animation: pulse 5s infinite;
+  -webkit-animation: pulse 5s infinite;
+  animation: pulse 5s infinite;
+  cursor: pointer;
+  transition: transform 0.2s;
+}
+
+.button-start:hover {
+  transform: scale(1.2);
+  animation: none;
```
Comment (trangtara) [inline code/style.css:91]: Pro!!!!

File: code/style.css [hunk 13]
```diff
@@ -1,58 +1,377 @@
...
+.game-wrapper {
+  flex-direction: column;
+  justify-content: space-evenly;
+  align-items: center;
   width: 90%;
-  height: 600px;
-  background: #ccc;
+  margin: 0 auto;
+  text-align: center;
+}
```
Comment (trangtara) [inline code/style.css:103]: Should we add display: flex here? I think so, but I will ask you seperately then :.)

File: code/style.css [hunk 14]
```diff
@@ -1,58 +1,377 @@
...
   width: 250px;
   height: 400px;
-  background: #000;
   text-align: center;
   padding-top: 30px;
+  padding-bottom: 20px;
   border-radius: 16px;
+  cursor: pointer;
+  background: #000;
+  margin: 25px 0;
```
Comment (trangtara) [inline code/style.css:141]: I like the way to collect similar styling of the decks into one class. Brilliant!

File: code/style.css [hunk 15]
```diff
@@ -1,58 +1,377 @@
...
+  margin-top: 100px;
+  width: 70px;
+  height: 70px;
+}
+
+.loader-container-animation {
+  -webkit-animation: scale-spin 1.5s linear infinite;
+  -moz-animation: scale-spin 1.5s linear infinite;
+  animation: scale-spin 1.5s linear infinite;
+}
```
Comment (trangtara) [inline code/style.css:174]: Good stuff!!! 👍

File: code/style.css [hunk 16]
```diff
@@ -1,58 +1,377 @@
...
+
+.reveal-wrapper {
+  flex-direction: column;
+  align-items: center;
+  width: 90%;
+  height: 100vh;
+  margin: 0 auto;
   text-align: center;
-  padding-top: 30px;
-  border-radius: 16px;
```
Comment (trangtara) [inline code/style.css:37]: Maybe I should ask myself instead "Do I need to have display: flex to make flex-direction: column work?" Huhm...I need to research more on this.

File: code/style.css [hunk 17]
```diff
@@ -1,58 +1,377 @@
...
   border-radius: 8px;
-  display: block;
-  margin: 0 auto;
+  margin: 50px auto;
+  font-family: 'Courier New', Courier, monospace;
+}
+
+.button:hover {
+  color: #fff;
+  transform: scale(1.1);
```
Comment (trangtara) [inline code/style.css:226]: C'est bien!

File: code/style.css [hunk 18]
```diff
@@ -1,58 +1,377 @@
...
+  }
+
+  .cards-container {
+    flex-direction: row;
+    flex-wrap: wrap;
+    justify-content: space-between;
+  }
+
+  .decks {
+    flex-basis: 30%;
```
Comment (trangtara) [inline code/style.css:248]: bravo!!!

File: code/style.css [hunk 19]
```diff
@@ -1,58 +1,377 @@
...
+  0% {
+    background-color: #000;
+    color: rgba(255, 255, 255, 0);
+    transform: scale(0.25);
+  }
+  75% {
+    background-color: crimson;
+    color: rgba(255, 255, 255, 1);
+    transform: scale(1);
+  }
```
Comment (trangtara) [inline code/style.css:371]: You have vey advanced animation effects on your page. Good job!

File: code/clue.js [hunk 20]
```diff
@@ -1,59 +1,454 @@
...
-  lastName: 'Green',
-  color: 'green',
-  description: 'He has a lot of connections',
-  age: 45,
-  image: 'assets/green.png',
-  occupation: 'Entrepreneur'
-}
+let killerPicked = false;
+let weaponPicked = false;
+let roomPicked = false;
```
Comment (trangtara) [inline code/clue.js:5]: This is smart!!!

File: code/clue.js [hunk 21]
```diff
@@ -1,59 +1,454 @@
...
+        'killerOccupation'
+      ).innerHTML = `Occupation: ${mystery.killer.occupation}`;
+      get('killerImage').src = mystery.killer.image;
+      get(
+        'killerDescription'
+      ).innerHTML = `Description: ${mystery.killer.description}`;
+      get(
+        'killerFavouriteWeapon'
+      ).innerHTML = `Weapon of choice: ${capitalizeFirstLetter(
+        mystery.killer.favourite
```
Comment (trangtara) [inline code/clue.js:234]: A very nice touch when you capitalize the first letters.

File: code/clue.js [hunk 22]
```diff
@@ -1,59 +1,454 @@
...
+  hideButton();
+  document.getElementsByTagName('BODY')[0].style.background = '#333';
+  shuffleFavourite(suspects, weapons);
+  shuffleFavourite(weapons, rooms);
+};
+
+/* The function is invoked when the 'killer card' deck is clicked. It checks that
+ a killer hasn't already been picked, then toggles the loader animation. After 
+ 3 seconds a killer is randomly selected and added to the mystery object and 
+ then displayed  */
```
Comment (trangtara) [inline code/clue.js:212]: I like the way you have a comment to explain how the function works. This makes me easier to understand your code. Good learning point.

File: code/clue.js [hunk 23]
```diff
@@ -1,59 +1,454 @@
...
+    get('mysteryKillerImage').src = mystery.killer.image;
+    get('mysteryRoom').innerHTML = `Room: ${capitalizeFirstLetter(
+      mystery.room.name
+    )}`;
+    get('mysteryWeapon').innerHTML = `Weapon: ${capitalizeFirstLetter(
+      mystery.weapon.name
+    )}`;
+  }
+};
+
```
Comment (trangtara) [inline code/clue.js:311]: I love this!

File: code/clue.js [hunk 24]
```diff
@@ -1,59 +1,454 @@
...
+const restartGame = () => {
+  location.reload();
+};
```
Comment (trangtara) [inline code/clue.js:314]: So nice!

File: code/clue.js [hunk 25]
```diff
@@ -1,59 +1,454 @@
...
+
+/* Increases the chance of the referred obeject's favourite property to be picked */
+const increaseChance = (object) => {
+  if (object === mystery.killer) {
+    arrayDuplicated = duplicateArray(findFavourite(mystery.killer, weapons));
+    return randomFavourite(weapons, arrayDuplicated);
+  } else if (object === mystery.weapon) {
+    arrayDuplicated = duplicateArray(findFavourite(mystery.weapon, rooms));
+    return randomFavourite(rooms, arrayDuplicated);
```
Comment (trangtara) [inline code/clue.js:364]: This is really advanced. nice job!

File: code/clue.js [hunk 26]
```diff
@@ -1,59 +1,454 @@
...
+  return favourite;
+};
+
+/* The function creates a new array where the favourite object occurs a random
+ number of times */
+const duplicateArray = (favourite) => {
+  const randomNumber = Math.floor(Math.random() * 11);
+  const arrayDuplicated = new Array(randomNumber).fill(favourite);
+  return arrayDuplicated;
+};
```
Comment (trangtara) [inline code/clue.js:383]: Very good!

File: code/clue.js [hunk 27]
```diff
@@ -1,59 +1,454 @@
...
+  } else if (elementId === 'loaderWeapon') {
+    pickWeapon(elementId);
+  } else if (elementId === 'loaderRoom') {
+    pickRoom(elementId);
+  }
+};
 
-// CREATE AN OBJECT THAT KEEPS THE MYSTERY.
-// With a killer, a weapon and a room.
-// The values will be set later.
```
Comment (trangtara) [inline code/clue.js:42]: remove unnecessary comment to keep the code clean

File: code/clue.js [hunk 28]
```diff
@@ -1,59 +1,454 @@
...
-// CREATE FUNCTIONS pickWeapon and pickRoom in a similar way.
+/* Event listener attached to the 'Reveal the crime' button. Calls the
+ revealMystery function */
+const revealButton = get('revealButton');
+get('revealButton').addEventListener('click', () => {
+  revealMystery(revealButton.parentNode.id);
+});
 
-// STEP 4 - CREATE A FUNCTION revealMystery that will be invoked when you click that button. It should show something like:
-// 'The murder was committed by Jacob Green, in the living room with a rope.'
```
Comment (trangtara) [inline code/clue.js:59]: remove these comment to clean the code

**Review by trangtara — COMMENTED** [review]
It's a very nice experience to review your code, Sara. 
You structured your code in a very friendly, easy-to-ready way. I just have some minor idea for improvement. Other than that, most of the time I see myself clapping hands and mouth-droping on your great work. 
I also learned a lot from you. So, thank you!

File: code/index.html (hunk 4 above)
Comment (sarahejhej) [reply code/index.html:17]: Haha, I just made the same comment to Ylva's code!

File: code/clue.js (hunk 27 above)
Comment (sarahejhej) [reply code/clue.js:42]: The red lines are already removed from the file. You see the - sign before the code :)

**Review by trangtara — APPROVED** [review]

File: code/clue.js [hunk 29]
```diff
@@ -1,59 +1,454 @@
...
+    id: 'patio',
+  },
+];
 
-// NOW GROUP ALL SUSPECTS, WEAPONS AND ROOMS IN ARRAYS LIKE THIS:
+const mystery = {
+  killer: '',
+  weapon: '',
+  room: '',
+};
```
Comment (HIPPIEKICK) [inline code/clue.js:194]: You could also declare them with `null`

File: code/clue.js [hunk 30]
```diff
@@ -1,59 +1,454 @@
...
+
+/* Called when the 'Let's Play' button is clicked, toggles the window, hides
+ the Reveal button and shuffles the suspects' favourite weapon and the weapons'
+ favourite room */
+const startGame = (windowId) => {
+  toggleWindow(windowId);
+  hideButton();
+  document.getElementsByTagName('BODY')[0].style.background = '#333';
+  shuffleFavourite(suspects, weapons);
+  shuffleFavourite(weapons, rooms);
```
Comment (HIPPIEKICK) [inline code/clue.js:206]: You should declare variables and functions before using them

File: code/clue.js [hunk 31]
```diff
@@ -1,59 +1,454 @@
...
+
+/* The function is invoked when the 'killer card' deck is clicked. It checks that
+ a killer hasn't already been picked, then toggles the loader animation. After 
+ 3 seconds a killer is randomly selected and added to the mystery object and 
+ then displayed  */
+const pickKiller = (loaderId) => {
+  if (!killerPicked) {
+    toggleAnimation(loaderId);
+    get('info').innerHTML = '';
+    setTimeout(function () {
```
Comment (HIPPIEKICK) [inline code/clue.js:217]:
```
setTimeout(() => {
  //code
}, 1500);
```

File: code/clue.js [hunk 32]
```diff
@@ -1,59 +1,454 @@
...
+ pickWeapon functions */
+const pickRoom = (loaderId) => {
+  if (weaponPicked && !roomPicked) {
+    toggleAnimation(loaderId);
+    get('info').innerHTML = '';
+    setTimeout(function () {
+      toggleLoader(loaderId);
+      mystery.room = increaseChance(mystery.weapon);
+      style('roomCard').background = mystery.killer.color;
+      get('roomName').innerHTML = `${capitalizeFirstLetter(mystery.room.name)}`;
```
Comment (HIPPIEKICK) [inline code/clue.js:278]: If it's only JavaScript, you don't need the template literals:
```
capitalizeFirstLetter(mystery.room.name);
```

File: code/clue.js [hunk 33]
```diff
@@ -1,59 +1,454 @@
...
+/* Capitalizes the first letter */
+const capitalizeFirstLetter = (inputToCheck) => {
+  return inputToCheck[0].toUpperCase() + inputToCheck.slice(1);
+};
 
-const weapons = []
+/* Function that returns an object to which the id refers */
+const get = (id) => {
+  return document.getElementById(id);
+};
```
Comment (HIPPIEKICK) [inline code/clue.js:410]: Maybe a more descriptive function name here would be "getElement" or "getDocument"

File: code/clue.js [hunk 34]
```diff
@@ -1,59 +1,454 @@
...
 
-const rooms = []
+/* Function that returns the style property of the referred object */
+const style = (id) => {
+  return get(id).style;
+};
```
Comment (HIPPIEKICK) [inline code/clue.js:415]: Naming suggestion: "styleElement"

File: code/clue.js [hunk 35]
```diff
@@ -1,59 +1,454 @@
...
+
+/* Hides the reveal the crime button when the game is started */
+const hideButton = () => {
+  const button = (style('revealButton').visibility = 'hidden');
+};
+
+/* Increases the chance of the referred obeject's favourite property to be picked */
+const increaseChance = (object) => {
+  if (object === mystery.killer) {
```
Comment (HIPPIEKICK) [inline code/clue.js:359]: We usually store an `id` property in objects to compare those values instead of the whole objects

**Review by HIPPIEKICK — APPROVED** [review]
Wow Sara, some advanced code you've got here 🤩 Keep it up with the descriptive comments and keep in mind to create really descriptive variable names and function names as well. 

This project _could_ be done in a simpler way (simpler does not equal better though). With that said, I think it's great that you're creating your own functions and try things out. It will be a valuable skill as we go forward with more complex projects.

@trangtara Keep it up with the nice comments 👍  For the future - it's totally fine to ask some questions about the code or to give suggestions for other ways of writing (if you've done it in a different way).

**Pull request closed without merging**